### ISSUE TITLE
Failed to call method "__get__" for descriptor class

### ISSUE THREAD

**Issue by yanyongyu** [microsoft/pyright#6452]
**Describe the bug**

Pyright failed to infer descriptor return type when using `self.xxx` in the class method. As shown in the following code, when using the descriptor outside, it's ok. But, failed when using it in the instance method.

The pyright says `Failed to call method "__get__" for descriptor class "API[xxx, xxx, xxx]"`.

Note: This is ok when using pyright 1.1.331, fail when using pyright 1.1.336

**Code or Screenshots**

Example:

[playground](https://pyright-play.net/?code=GYJw9gtgBMCuB2BjALmMAbAzlAlhADmCMlPgIbE5noBQokUyAnvjvAOa4FEkDiApvH4gciADRQAKi34A1ChIAKFMhADK%2BfuKgBhMEjLJBh-hJ3V0ZAEbpTUAIIB3MjmTXbEgPL5kOfdQlpTQkwADdhdDAyABMaGgAhKABeKRl5EAAKACJ4rIBKGgAlZNTNdOzC-JpFEuUQVQ0tbMUquMRLTGx7RQBJDIEhEUQAbXilCUKAXTyALhooBaho-mAoAH01tlcNjMx%2BdGAJOCQZ3Qt3fmG9AyN4E1GlSYknFzcbS6npqABaAD4oABy%2Bn4c0WYKgewOADpjogSrC4mDlqsNntkGs7hB%2BDtIYcoGBHINTkFLvEnlBMSCIcgQHkfv8gUJQeCFrioZSSpTEYsAAJhCJRWJIlbrNbsfjota7fZ4sBWABWp0Zdjl8uYmmJMlGXz%2BUCy3R6DygigmkyyzJZUKt3IWfPCIEiMXmi2RovFkulBxCCtOY3xCvVVO8vn86GGJO1Ov%2B5nQlnewxNDmcrguw0%2BFvBVqhNqWIo27p2zpZuO9iqgwb8dzDZNLgdOFdD4a1ZMmJWVRagdN1%2Bt6RsTUygAB8zrHU4mXin459zR2wThVqrcNhlRmWQsQBLYCB4BCZTmwRvkFud%2BRKNRPcAYQhtKq6VAAMSMGSnHDseBEfhxGjtMidBxzDtfyYJBc1WIxMGQC9ZlnRYN3Cag1kDC8ryQAo113aFYQyABGCQrCSLIACZWjBHkDUAzBgLhV0sJLKAyBfeBkAkAAqPDTgg2l6WpEBV0WLMvzIEp7AyAo4P4BCkLIFDEAKaSsNwqB8KIqogA)

```python
from functools import partial
from typing import Generic, TypeVar, ParamSpec, Concatenate, Callable, Awaitable, Optional, Type, overload

B = TypeVar("B")
R = TypeVar("R")
P = ParamSpec("P")


class API(Generic[B, P, R]):
    def __init__(self, func: Callable[Concatenate[B, P], Awaitable[R]]) -> None:
        self.func = func

    def __set_name__(self, owner: Type[B], name: str) -> None:
        self.name = name

    @overload
    def __get__(self, obj: None, objtype: Type[B]) -> "API[B, P, R]":
        ...

    @overload
    def __get__(self, obj: B, objtype: Optional[Type[B]]) -> Callable[P, Awaitable[R]]:
        ...

    def __get__(
        self, obj: Optional[B], objtype: Optional[Type[B]] = None
    ) -> "API[B, P, R] | Callable[P, Awaitable[R]]":
        if obj is None:
            return self

        return partial(self.func, obj)  # type: ignore


class A:

    async def test(self):
        reveal_type(self.func)  # Error
        self.func(1, b="2")

    @API
    async def func(self, a: int, *, b: str) -> str:
        ...


a = A()
reveal_type(a.func)  # OK
a.func(1, b="2")

```

**VS Code extension or command-line**

Pyright 1.1.336, command-line

**Comment by erictraut** [comment]
Thanks for the bug report. This was a long-standing bug that was exposed by a recent change I made.

I've fixed the bug and incorporated a variation of your code sample into pyright's unit tests to prevent a regression in the future. The fix will be included in the next release.

**Comment by medihack** [comment]
Not sure if this is the same bug, but at least the same error message ;-) I am using the VS Code extension with v1.336, but it also failed with v1.335. The last working version here was v1.334.
As a side note, the published v1.336 isn't on PyPI yet.

```python
class BaseDataset:
    def __init__(self):
        self._modality = "MR"

    @property
    def modality(self) -> str:
        return self._modality


class QueryDataset(BaseDataset):
    @BaseDataset.modality.setter
    def modality(self, value: str) -> None:
        self._modality = value


ds = QueryDataset()
x = ds.modality # <-- similar error
```

**Comment by erictraut** [comment]
@medihack, that's not the same issue. Please file a separate bug report if you'd like me to take a look at that one.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
